packages/vite/src/node/server/moduleGraph.ts

```diff
@@ -238,7 +238,7 @@ export class ModuleGraph {
     const resolved = await this.resolveId(url, !!ssr)
     const resolvedId = resolved?.id || url
     const ext = extname(cleanUrl(resolvedId))
-    const { pathname, search, hash } = parseUrl(url)
+    const { pathname, search, hash } = new URL(url, 'relative://')
```

Comment:
I think this changes the behavior of Line 243.
```shell
$ node -e "console.log(require('url').parse('\0virtual:aa'))" # node 16.14.1
Url {
  protocol: null,
  slashes: null,
  auth: null,
  host: null,
  port: null,
  hostname: null,
  hash: null,
  search: null,
  query: null,
  pathname: '\x00virtual:aa',
  path: '\x00virtual:aa',
  href: '\x00virtual:aa'
}
$ node -e "console.log(require('url').parse('\0virtual:aa'))" # node 16.16.0
Url {
  protocol: 'virtual:',
  slashes: null,
  auth: null,
  host: 'aa',
  port: null,
  hostname: 'aa',
  hash: null,
  search: null,
  query: null,
  pathname: null,
  path: null,
  href: 'virtual:aa'
}
$ node -e "console.log(new URL('\0virtual:aa', 'relative://'))" # node 16.16.0
URL {
  href: 'virtual:aa',
  origin: 'null',
  protocol: 'virtual:',
  username: '',
  password: '',
  host: '',
  hostname: '',
  port: '',
  pathname: 'aa',
  search: '',
  searchParams: URLSearchParams {},
  hash: ''
}
```

```js
'\x00virtual:aa' // url.parse with node 16.14.1
'aa' // new URL with node 16.16.0
```

Reply:
please try this instead: `node -e "const { URL }=require('url'); console.log(new URL('\0virtual:aa', 'relative://'))"`

Reply:
Yes, it does. But it has a pathname now, while url.parse has `null`.

Reply:
Do we need the pathname and thus the returned `url` to be the same as before? As far as I've seen it's used as a key to `urlToModuleMap`, which should work inside of `ModuleGraph`.

However, it's also used in import-analysis plugin and `urlToModuleMap` is read from in ssrStacktrace, which might be problematic. Tests don't seem to cover that, though - they're all green.

Reply:
> Do we need the pathname and thus the returned `url` to be the same as before?

I'm not familiar to code around here enough to say no. But I suppose it will break things when multiple urls got the same key for `urlToModuleMap`.

Reply:
Do you known who could help? I’m surprised this is not a big issue to projects building with vite yet 🙈

Reply:
I'll review it in the next days. We'll release 3.0.1 with current fixes to main tomorrow, and then review this issue to resolve it in 3.0.2. Thanks for digging into this @codepunkt

Reply:
You're welcome. Thanks to you, @sapphi-red and all other regular maintainers!
If there's anything I can do to help, let me know!